core/unix/os.c

```diff
@@ -8363,10 +8349,25 @@ process_mmap(dcontext_t *dcontext, app_pc base, size_t size, uint prot,
              */
             LOG(THREAD, LOG_SYSCALLS, 4, "mmap " PFX ": partial\n", base);
         } else {
+#ifdef ANDROID
+            /* The Android loader's initial all-segment-covering mmap is anonymous */
+            dr_mem_info_t info;
+            if (query_memory_ex_from_os((byte *)ALIGN_FORWARD(base + size, PAGE_SIZE),
+                                        &info) &&
+                info.prot == MEMPROT_NONE && info.type == DR_MEMTYPE_DATA) {
+                LOG(THREAD, LOG_SYSCALLS, 4, "mmap " PFX ": Android elf\n", base);
+                image = true;
+                DODEBUG({ map_type = "ELF SO"; });
+                os_add_new_app_module(dcontext, true /*at_map*/, base,
+                                      /* pass segment size, not whole module size */
+                                      size, memprot);
+            }
+#else
```

Comment:
This code below should be enabled for Android as well for the per-segment mappings, just like in the original code.  There should be an `else` for this for Android, just like before.

Reply:
Again for the future: please mark each suggestion as "Done" if addressed.